### PR TITLE
Calculator - Guided Tutorial Complete

### PR DESCRIPTION
**Pull request by MajorPayne95**
[I reworked a lot of the Microsoft tutorial code in an effort to make the main loop easier to read, and to work on OOP principles.

### CODE REVIEW

File: Calculator.MajorPayne95/Calculator/Program.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
+using CalculatorLibrary; // Ensure this is included to access the CalculatorLibrary namespace
+using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+
+
+namespace CalculatorProgram
+{
```
Comment (nwdorian) [inline Calculator.MajorPayne95/Calculator/Program.cs:9]: 🟠 **Namespaces**  

❓As per the academy's [code-conventions](https://thecsharpacademy.com/article/58/code-conventions), use file-scoped namespaces.  

➡️ From C# 10 we are able to remove the namespace block and save one indentation level, which optimizes the use of space and improves readability.

💡 If you're using Visual Studio 2022, here's a short article to setup [file-scoped namespaces by default](https://darthpedro.net/2024/01/29/configure-visual-studio-to-use-file-scoped-namespaces-on-new-classes/).
```suggestion
namespace CalculatorProgram;
```

File: Calculator.MajorPayne95/Calculator/Program.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
+using CalculatorLibrary; // Ensure this is included to access the CalculatorLibrary namespace
+using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
```
Comment (nwdorian) [inline Calculator.MajorPayne95/Calculator/Program.cs:5]: 🟠**Unnecessary using directives**

➡️ Remove anything that's not being used. It makes your code less readable and maintainable. 

💡In Visual Studio the compiler will tell you when code isn't being used by slightly fading it. Here's a short video on how to make Visual Studio [automatically remove unused using statements](https://www.youtube.com/watch?v=CC5nohNp7F8)

image

File: Calculator.MajorPayne95/Calculator/Program.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            int usageCount = 0; // Initialize usage count
+
+            List<string> history = new List<string>();
+
+            Calculator calculator = new Calculator();
+            UsageLogger usageLogger = new UsageLogger();
+            UIManager uiManager = new UIManager();
+
+            // Display title as the C# console calculator app.
+            uiManager.DisplayTitle();
```
Comment (nwdorian) [inline Calculator.MajorPayne95/Calculator/Program.cs:25]: 🟡 **Code comments**

👉 It's fine to add comments during development if it's going to help you with pseudo-code or to remember some specific logic. When you finish a project, always go back through the code for a polish pass and clean up unnecessary comments.  

💭 Remember, our code should be self-documenting - with good naming and structure we shouldn't need any comments at all.

➡️ For example, this method has good name and we instantly know what it does, there is no need for the comment exaplaining that it displays the title.

**Review by nwdorian — APPROVED** [review]
Hey @MajorPayne95 👋,

Excellent work on your **Calculator** project submission! 🎉

I have finished the code review and left some feedback below. ☺️

Feel free to reach out if you have questions or if you'd like to collaborate on any of these points.✌️

---

🟢 **Requirements**

⭐ You have completed all of the project requirements!

---

🏅 **Extra challenges**

✅ Count the amount of times the calculator was used
✅ List of latest calculations
✅ Use previous results for new calculations
✅ Add extra operations
💪 Very nice to see you implemented extra challenges as it speeds up the learning process! 

---

🟠 **Missing .gitignore**

❓ It's important to have a `.gitignore` file in order to prevent adding unnecessary folders and files to source control (like `bin` and `obj` folders). You can find a template specific to .NET [in this repo](https://github.com/github/gitignore/blob/main/Dotnet.gitignore) or add it with a `dotnet new gitignore` CLI command.

💡After initializing the git repository for a new project, add  `.gitignore` and add a commit with "initial commit" message. Embracing this practice will make sure you never commit unnecessary stuff. 

---

🟡 **Multiple classes in the same file** 

❓In general we want to have one class per file. This makes it easier to navigate code for others, whether they're peers, mentors, or even future employers. 

➡️ Feel free to use folders as much as you want to organize classes into logical groups. 📁

---

🟡 **Missing 2nd project in the solution**

❓When I open the solution I can't see your `CalculatorLibrary` project. If you have multiple projects per application, make sure you add them to the solution. 

image

---

🟡 **Application flow bug**

❓When asked to use current result for a new calculation and selecting "no", there is no prompt for first number, resulting in invalid input message no matter what I put in. 

➡️ Make sure to thoroughly test the application before submitting for review. Cover happy and failure paths, invalid inputs and edge cases. Code reviews get stricter in future projects and you might be asked to fix issues like this.

image

image

---

💪 I love that you used multiple classes to practice object oriented principles. You must have learned a lot by taking the code a step further and I'd like to see you advance to the next step in the academy roadmap. I'm sure future projects will be even better with all the new knowledge. 

💭 You can take care of issues in your own repo if you want, but since the project will be approved you don't have to push the changes or create a new PR.

✅ I will go ahead and approve the project submission. Keep up the good work on the next ones! 😄

Best regards,  
@nwdorian
